### PR TITLE
beaver install: Support multiple Dockerfile contexts

### PR DESCRIPTION
**Pull request by leandro-lucarella-sociomantic**
Sometimes is useful to have multiple different docker contexts for a project. The most typical case is having one (`beaver.`)`Dockerfile` for building a project (with the build dependencies), and another one to run the project (with the runtime dependencies). Since Docker doesn't support specifying a `.dockerignore` file yet, we overcome this limitation in beaver.

To use multiple build contexts, you can specify the variable `BEAVER_DOCKER_CONTEXT` to point to a directory that will contain your `Dockerfile` or `beaver.Dockerfile` and possibly a `build` script (equivalent to the `docker/build` script in normal builds) and/or a `dockerignore` (`.dockerignore is also accepted but `dockerignore` will take precedence), which will be copied to the current working directory as `.dockerignore`.

For example:
```
docker/builder/beaver.Dockerfile
docker/builder/build
docker/builder/dockerignore

docker/runner/Dockerfile.xenial
docker/runner/dockerignore
```

Where:

```console
$ cat docker/builder/dockerignore
*
!docker/builder/

$ cat docker/runner/dockerignore
*
!docker/runner/
```

(will ignore all files except for its context, all paths should be relative to the current working directory, same for the `Dockerfile`s)

### CODE REVIEW

**Comment by leandro-lucarella-sociomantic** [comment]
You can also look at the tests for a concrete example. The usage is still ugly, but I think we can CLI args support on top of it.

File: bin/beaver-install [hunk 1]
```diff
@@ -7,13 +7,35 @@ set -eu
 
 d="$(dirname "$0")"
 
+# Copy a file only if it exist
+maybe_cp()
```
Comment (nemanja-boric-sociomantic) [inline bin/beaver-install]: :giggle:

**Comment by leandro-lucarella-sociomantic** [comment]
I had to change the use of v4 cachalot image with v2 as the rest of the tests because this is testing trusty too, and there are no v4 images for trusty. At some point I guess we should remove trusty traces from beaver :)

**Comment by leandro-lucarella-sociomantic** [comment]
Actually I could have changed the `DIST` to use `xenial` (so now it would work with v4), but still, maybe is better to update all the images used in tests in one PR.

File: bin/beaver-install [hunk 2]
```diff
@@ -7,13 +7,35 @@ set -eu
...
+then
+    maybe_cp "$BEAVER_DOCKER_CONTEXT/.dockerignore" .dockerignore
+    maybe_cp "$BEAVER_DOCKER_CONTEXT/dockerignore" .dockerignore
+    docker_dir="-d $BEAVER_DOCKER_CONTEXT"
+else
+    BEAVER_DOCKER_CONTEXT=.
+    docker_dir=
+fi
+
 if test -r "Dockerfile${DIST:+.$DIST}"
```
Comment (mihails-strasuns-sociomantic) [inline bin/beaver-install]: Missing `BEAVER_DOCKER_CONTEXT`

**Review by mihails-strasuns-sociomantic — COMMENTED** [review]
Going to try this with dhtnode. Looks good setting aside CLI/env issue we talked about before.

**Comment by mihails-strasuns-sociomantic** [comment]
Tried with dhtnode repo, this command `BEAVER_DOCKER_IMG=builder DMD=2.070.2.s* F=production DIST=xenial BEAVER_DOCKER_CONTEXT=docker/builder beaver dlang install` does not catch build script `docker/builder/build`.

**Comment by leandro-lucarella-sociomantic** [comment]
That's weird, tests seem to work well.

File: bin/beaver-install (hunk 2 above)
Comment (leandro-lucarella-sociomantic) [reply bin/beaver-install]: Added a fixup commit for this. Good catch!

Comment (leandro-lucarella-sociomantic) [reply bin/beaver-install]: (I was about to add a test for that before and said "meh, who uses that anyway :P")

**Comment by leandro-lucarella-sociomantic** [comment]
Can you maybe add a print line for `$docker_dir/$docker_build` here:
https://github.com/sociomantic-tsunami/beaver/blob/eebf4a3b9ee7941e47ec0ce29eb0ca6cb71b9681/bin/docker/gen-dockerfile#L118

And post the output of running it?

**Comment by leandro-lucarella-sociomantic** [comment]
Oh, I just see this is `dlang install`, maybe there is some trickery there... I'll check!

**Comment by leandro-lucarella-sociomantic** [comment]
Ah, yeah, it does its own `gen-dockerfile`. I thought I searched for all the uses of it. I'll fix.

**Comment by leandro-lucarella-sociomantic** [comment]
Well, it actually does its own complete thing, it doesn't user `beaver install` at all :roll_eyes: (for now I will just fix it, and afterwards we can, with improved CLI support, use `beaver install` instead).

**Comment by mihails-strasuns-sociomantic** [comment]
Yeah, I think this is exactly how I ended up implementing forwarding of `-I` flags in one of previous PR drafts. Right now `dlang install` and `install` are completely duplicated.

Red btw ;)

**Comment by leandro-lucarella-sociomantic** [comment]
Finally tests pass, and I added support to `dlang install` too, for now copy&past, but I prefer to have something usable first and then beautify it. :)

Have a final look and then I can fixup and merge once I get the LGTMs

**Comment by mihails-strasuns-sociomantic** [comment]
Yup, testing right now

**Comment by mihails-strasuns-sociomantic** [comment]
Yup, works for builder but for runner I got this:

```
~/core/dhtnode [dockerrun] [v1.11.0-9-g3c0de60]
✔ 10:44:34 $ BEAVER_DOCKER_IMG=dhtnode DMD=2.070.2.s* F=production DIST=xenial BEAVER_DOCKER_CONTEXT=docker/runner beaver install
cat: docker/runner/beaver.Dockerfile: No such file or directory
~/core/dhtnode [dockerrun] [v1.11.0-9-g3c0de60]
✖ 1 10:44:58 $ ls docker/runner/
Dockerfile
```

**Comment by mihails-strasuns-sociomantic** [comment]
Ah, right, it has to be `Dockerfile.xenial` if `DIST` is defined

**Comment by mihails-strasuns-sociomantic** [comment]
Should probably remove current `.dockerignore` if context dir doesn't define one

**Comment by leandro-lucarella-sociomantic** [comment]
> Should probably remove current .dockerignore if context dir doesn't define one

Ah, good catch. In general the tool doesn't remove anything for debugging purposes, but maybe we should remove files by default after it's built and add a way to keep them as an option for debugging/tests.

**Comment by leandro-lucarella-sociomantic** [comment]
(same with `beaver.Dockerfile.generated`)

**Comment by leandro-lucarella-sociomantic** [comment]
OK, updated with that addition.

**Comment by mihails-strasuns-sociomantic** [comment]
Seems to work. I have updated dhtnode PR in a moment so that you can see how it looks for yourself (https://github.com/sociomantic-tsunami/dhtnode/pull/58)

**Comment by leandro-lucarella-sociomantic** [comment]
RED? :grumpy:

**Comment by leandro-lucarella-sociomantic** [comment]
Dammit, it works locally :(

**Comment by leandro-lucarella-sociomantic** [comment]
OK, it should be green in a few minutes :crossed_fingers:

**Review by mihails-strasuns-sociomantic — APPROVED** [review]
You intend to make CLI stuff as a separate PR, right?

**Comment by leandro-lucarella-sociomantic** [comment]
> You intend to make CLI stuff as a separate PR, right?

Yup

**Comment by leandro-lucarella-sociomantic** [comment]
Feel free to release v0.5.0 with this. I won't be able to work on CLI improvement in the next 3 weeks anyway.

**Pull request merged**